openfl/federated/data/sources/pytorch/verifiable_map_style_dataset.py

```diff
@@ -0,0 +1,49 @@
+# Copyright 2020-2025 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from abc import abstractmethod
+
+import numpy as np
+import torch
+
+from openfl.federated.data.sources.verifiable_dataset_info import (
+    VerifiableDatasetInfo,
+)
+
+
+class VerifiableMapStyleDataset(torch.utils.data.Dataset):
+    """Base class for different types of map style datasets."""
+
+    def __init__(self, vds: VerifiableDatasetInfo, transform=None, verify_dataset=False):
+        self.transform = transform
+        self.verifiable_dataset_info = vds
+        self.verify_dataset = verify_dataset
+        self.datasets = self.create_datasets()
+
+        # create indices for fast lookup
+        lengths = list(map(len, self.datasets))
+        self.cumulative_sizes = np.cumsum(lengths)
+
+    def __getitem__(self, idx):
+        # find which sub-dataset this index belongs to
+        dataset_idx = self.cumulative_sizes.searchsorted(idx, side="right")
+        # find the data in that sub-dataset
+        data_idx = idx - self.cumulative_sizes[dataset_idx - 1] if dataset_idx > 0 else idx
+        item = self.datasets[dataset_idx][data_idx]
+        data_path = item["path"]
+
+        if self.verify_dataset:
+            item_hash = self.verifiable_dataset_info.data_sources[dataset_idx].compute_file_hash(
+                data_path
+            )
+            if not self.verifiable_dataset_info.verify_single_file(data_path, item_hash):
+                raise ValueError(f"Data integrity check failed for {data_path}")
```

Comment:
@vrancurel, @Efrat1, I'm wondering if this `verify_dataset` flag is actually useful... If the dataset doesn't need to be verified, then why not simply instantiate a `torch.utils.data.Dataset`? On the other hand, load-time verification would be enforced for the `VerifiableMapStyleDataset`. WDYT?

Reply:
Well, an `if` statement in a method implementation is sometimes (often) a "smell" of some problem with the class hierarchy... Could you consider removing the flag which is confusing IMO and inherit directly from `ImageFolder` if per-item verification isn't required?

Reply:
Understood the purpose of the `self.verify_dataset` flag, and the fact that the `VerifiableMapStyleDataset` also adds the ability to use multiple data sources.